--- /usr/bin/debmirror	2009-08-29 19:28:59.000000000 +0200
+++ bin/debmirror_2.1_mine	2009-08-30 19:16:50.000000000 +0200
@@ -62,6 +62,7 @@
           [--di-dist=foo[,bar,..] ...] [--di-arch=foo[,bar,..] ...]
           [--source|--nosource] [--getcontents] [--md5sums]
           [--ignore-missing-release] [--ignore-release-gpg]
+          [--trash=dir]
           [--ignore=regex] [--exclude=regex] [--include=regex]
           [--exclude-deb-section=regex] [--limit-priority=regex]
           [--timeout=seconds] [--max-batch=number]
@@ -260,6 +261,10 @@
 
 Don't fail if the Release.gpg file is missing.
 
+=item --trash=dir
+
+When cleaning up, instead of deleting, move files in this directory.
+
 =item --ignore=regex
 
 Never delete any files whose filenames match the regex. May be used multiple times.
@@ -516,6 +521,8 @@
 use Digest::MD5;
 use Digest::SHA1;
 use LWP::UserAgent;
+use File::Basename;
+use File::Copy;
 
 # Yeah, I use too many global variables in this program.
 our ($debug, $progress, $verbose, $passive, $skippackages, $getcontents);
@@ -529,6 +536,7 @@
 our $cleanup=0;
 our $post_cleanup=1;
 our $no_cleanup=0;
+our $trash_dir=0;
 our $do_source=1;
 our $host="ftp.debian.org";
 our $user="anonymous";
@@ -608,6 +616,7 @@
 	   'rsync-extra=s'          => \@rsync_extra,
 	   'cleanup'                => \$cleanup,
 	   'postcleanup'            => \$post_cleanup,
+  	   'trash=s'                => \$trash_dir,
 	   'nocleanup'              => \$no_cleanup,
 	   'ignore=s'               => \@ignores,
 	   'exclude=s'              => \@excludes,
@@ -2065,7 +2074,14 @@
       if ($files{$file} == 2 && -f $file) {
 	say("deleting $file") if ($verbose);
 	if (! $dry_run) {
+         if ( ! $trash_dir ) { 
 	  unlink $file or die "unlink $file: $!";
+	 } else {
+             rename ($file, ($trash_dir . "/" . basename($file) )) or
+	     { copy ($file, ($trash_dir . "/" . basename($file) )) and
+                   unlink $file
+             }  or  die "cannot move or copy $file to trash: $!";
+	 }
 	}
       }
     }
