--- /usr/bin/debmirror	2006-06-25 18:22:32.000000000 +0200
+++ bin/debmirror.mine	2006-06-25 21:22:07.161553760 +0200
@@ -57,7 +57,7 @@
           [--timeout=seconds] [--root=directory]
           [--dist=foo[,bar,..] ...] [--section=foo[,bar,..] ...]
           [--arch=foo[,bar,..] ...] [--skippackages] [--getcontents]
-          [--cleanup|--nocleanup|--postcleanup] [--adddir=directory]
+          [--cleanup|--nocleanup|--postcleanup] [--trash=dir] [--adddir=directory]
           [--ignore=regex] [--exclude=regex] [--include=regex]
           [--exclude-deb-section=regex] [--max-batch=number]
           [--rsync-batch=number] [--ignore-missing-release]
@@ -178,6 +178,10 @@
 Clean up the local mirror but only after mirroring is complete and
 only if there was no error.
 
+=item --trash=dir
+
+When cleaning up, instead of deleting, move files in this directory.
+
 =item --ignore=regex
 
 Never delete any files whose filenames match the regex. May be used multiple times.
@@ -356,6 +360,7 @@
 use Compress::Zlib;
 use Digest::MD5;
 use LWP::UserAgent;
+use File::Basename;
 
 # Yeah, I use too many global variables in this program.
 our ($debug, $progress, $verbose, $passive, $skippackages, $getcontents);
@@ -366,6 +371,7 @@
 our $check_downloads = 0;
 our $cleanup=1;
 our $post_cleanup=0;
+our $trash_dir=0;
 our $do_source=1;
 our $host="ftp.debian.org";
 our $user="anonymous";
@@ -415,6 +421,7 @@
 	   'adddir=s'     => \@extra_dirs,
 	   'cleanup!'     => \$cleanup,
 	   'postcleanup!' => \$post_cleanup,
+	   'trash=s'      => \$trash_dir,
 	   'ignore=s'     => \@ignores,
 	   'exclude=s'    => \@excludes,
 	   'exclude-deb-section=s' => \@excludes_deb_section,
@@ -1471,7 +1478,11 @@
     unless (exists $files{$file} or (defined($ignore) && $file=~/$ignore/o)) {
       say("deleting $file") if ($verbose);
       if (! $dry_run) {
-       unlink $file or die "unlink $file: $!";
+	  if ( ! $trash_dir ) {
+	      unlink $file or die "unlink $file: $!";
+	  } else {
+	      rename $file, ($trash_dir . "/" . basename($file) ) or  die "rename $file: $!";
+	  }
       }
     }
   }
